---
title: "Distance Based Methods- group project"
author: "Dani Antos"
date: "November 7, 2017"
output: html_document
---

to make link: [whatever they said](http link)
table of contents: in HEADER, colon after html_document (in output), then toc:true, toc_depth: 4 (number of subsections that will be referenced in the table of contents), toc_float: true (it will be on the side, easier to access)
tabs: in header, include {.tabset} after header line (does it for the other levels, like if you "tab" a header, then the sections will be tabbed within the header) 
Install packages: {ape}, {phangorn}, {adegenet}

```{r read data into rstudio}
library(ape)
library(adegenet)
library(phangorn)
dna <- fasta2DNAbin(file="http://adegenet.r-forge.r-project.org/files/usflu.fasta")
dna
class(dna)
as.character(dna)[1:5,1:10]
unclass(dna)[1:5,1:10]
typeof(unclass(dna)[1:5,1:10])
object.size(as.character(dna))/object.size(dna)

annot <- read.csv("http://adegenet.r-forge.r-project.org/files/usflu.annot.csv", header=TRUE, row.names=1)
head(annot) #we have 80 isolates (5 per year) typed (?) for the same 1701 nucleotides
```

Distance-based methods: get a matrix of pairwise genetic distances between taxa and then use hierarchical clustering to create the tree

clustering methods:

1. UPGMA- simplest method for constructing trees, assumes the same evolutionary speed for all lineages (disadvantage); all leaves have the same distance from the root (creates ultrametric tree)

2. Neighbor-joining- taking the two closest nodes of the tree and defined as neighbors; you keep doing this until all of the nodes have been paired together

Advantages? It's fast and flexible

Disadvantages? Can't test for the best model/tree, can be inaccurate, and is highly dependent on the distance and clustering algorithm

Remember: Basic steps for Distance-Based Phylogenies

1. find genetic distances for pairs of individuals (in our case, isolates)

2. make a tree using these distances

3. evaluate the relevance of your tree

##Step 1
```{r step 1}
library(ape)
D <- dist.dna(dna, model = "TN93")
length(D) #number of pairwise distances, computed as n(n-1)/2
temp <- as.data.frame(as.matrix(D))
table.paint(temp, cleg=0, clabel.row=.5, clabel.col=.5) #darker shades of gray mean a larger distance # you can also make cool color plots but they're much more complicated because they use the image() function
#we can start to see a pattern because the data is ordered by year, but we can't really make any conclusions yet
```
New functions:

- dist.dna() from {ape} package: this literally just makes a matrix of pairwise distances from the DNA we give it, super easy; "TN93" is just the type of evolutionary model we're using, this particular one allows for different transition rates, heterogenous base frequencies, and variation of substitution rate at the same site

That was step 1 in its entirety, so it's pretty easy because we have a nice function that computes all of our distances for us.

##Step 2
![](http://quentincordonnier.tumblr.com/post/90647281369/growin-up)

Building the tree generally just gives us a better visual understanding of what's going on. As I mentioned previously, the figure from Step 1 shows us that there is a trend or pattern of some sort, but it's difficult to get more than that just from the figure. On the other hand, we can't always assume that a tree is the best or most efficient representation of our genetic distances.

R has a ton of algorithm options for us to choose from to make our tree (New functions):

- nj() from {ape} package: classic Neighbor-Joining algorithm

- bionj() from {ape} package: Neighbor-Joining 2.0 (basically)

- fastme.bal() AND fastme.ols() both from {ape} package: minimum evolution algorithm (to my understanding, usually looks the same as neighbor joining, topology shows the smallest value of branch sums)

- hclust() from {stats} package (a base package in R): classical hierarchical clustering, including UPGMA and others

```{r step 2}
tre <- nj(D)
class(tre) #all trees created using {ape} package will be of class phylo
tre <- ladderize(tre)
tre # tells us what the tree will look like but doesn't show the actual construction
plot(tre, cex = 0.6)
title("A Simple NJ Tree")
# or 
h_cluster <- hclust(D, method = "average", members = NULL) # method = average is used for UPGMA, members can be equal to NULL or a vector with a length of size D
plot(h_cluster, cex = 0.6)
```
Feel free to try the other three functions out on your own! How do they compare to these two examples? Do you think one is better than the rest?

The two previous examples show the simplest types of trees that we can make, but you can also add annotations, labels, and colors to make the tree easier to interpret. 

*EXAMPLE*: we can use colors to represent different years
```{r trees with color}
plot(tre, show.tip=FALSE) # gets rid of the labels on the end, refer to the first tree depicted above
title("Unrooted NJ tree")
myPal <- colorRampPalette(c("red","yellow","green","blue"))
tiplabels(annot$year, bg=num2col(annot$year, col.pal=myPal), cex=.5) #we use the annot dataset to get our years
temp <- pretty(1993:2008, 5)
legend("bottomleft", fill=num2col(temp, col.pal=myPal), leg=temp, ncol=2)
```
New functions:

- colorRampPalette() in {grDevices} package: takes a certain set of colors and returns new color palettes or color "ramps" to map an interval

- tiplabels() in {ape} package: adds labels to or near nodes/edges of a tree

- num2col() in {adegenet} package: can translate variables onto a color scale

- pretty(): computes a sequence of equally spaced values that encompass the range of x values

The tree is now easy to read with the color coding, but the location of some of the isolates can be deceiving due to the unrooted nature of the tree. We can re-draw the tree like this:
```{r redrawn unrooted tree}
plot(tre, type = "unrooted", show.tip = FALSE)
title("Unrooted NJ Tree")
tiplabels(tre$tip.label, bg = num2col(annot$year, col.pal = myPal), cex = 0.5)
```
Or, even better, we can make a root for the tree. The best rooting would be any of the oldest isolates. We can use the annot dataset again.
```{r rooting tree}
head(annot)
tre2 <- root(tre, out = 1)
tre2 <- ladderize(tre2)
plot(tre2, show.tip=FALSE, edge.width=2)
title("Rooted NJ tree")
tiplabels(tre$tip.label, bg=transp(num2col(annot$year, col.pal=myPal),.7), cex=.5, fg="transparent")
axisPhylo()
temp <- pretty(1993:2008, 5)
legend("topright", fill=transp(num2col(temp, col.pal=myPal),.7), leg=temp, ncol=2)
```
New  functions:

- root() fomr {ape} package: reroots a phylo tree using a specific outgroup

- axisPhylo() from {ape} package: adds a scaled axis on the side of a plot

**They go into molecular clocks here and how to perform a regression to figure out if there is a molecular clock, I might not do that- pg. 12 if I decide to add it in later**

##Step 3
Because there are so many different algorithms to choose from when constructing our tree, we have to make sure the one we chose was appropriate for our dataset, using our original distance matrix (in this case, D). This is much easier than it sounds, and just requires some plots and correlation calculations. 
```{r step 3}

```

